docusaurus/docs/React/components/message-components/attachment/voice-recording.mdx

```diff
@@ -0,0 +1,251 @@
+            alt: 'Image of the VoiceRecordingPlayer playing the audio',
+        },
+    ]}
+/>
+
+## UI Fallbacks
+
+### Missing duration
+
+If the duration is not available in the [attachment object payload](./#attachment-payload), `VoiceRecordingPlayer` as well as `QuotedVoiceRecording` component will display the attachment size instead.
+
+<ImageShowcase
+    border
+    items={[
+        {
+            image: VoiceRecordingPlayerFileSizeFallback,
+            caption: <span>VoiceRecordingPlayer displaying file size instead of audio duration</span>,
+            alt: 'Image of the VoiceRecordingPlayer displaying file size instead of audio duration',
+        },
+        {
+            image: QuotedVoiceRecordingFileSizeFallback,
+            caption: <span>QuotedVoiceRecording displaying file size instead of audio duration</span>,
+            alt: 'Image of the QuotedVoiceRecording displaying file size instead of audio duration',
+        },
+    ]}
+/>
+
+
+### Missing title
+
+If the voice recording does not come with title, a fallback title is provided.
+
+<ImageShowcase
+    border
+    items={[
+        {
+            image: VoiceRecordingPlayerFallbackTitle,
+            caption: <span>VoiceRecordingPlayer displaying the fallback title</span>,
+            alt: 'Image of the VoiceRecordingPlayer displaying the fallback title',
+        },
+        {
+            image: QuotedVoiceRecordingFallbackTitle,
+            caption: <span>QuotedVoiceRecording displaying the fallback title</span>,
+            alt: 'Image of the QuotedVoiceRecording displaying the fallback title',
+        },
+    ]}
+/>
+
+### Missing `waveform_data`
+
+If the `waveform_data` is an empty array, then no progress bar is rendered.
+
+<ImageShowcase
+    border
+    items={[
+        {
+            image: VoiceRecordingPlayerMissingWaveformData,
+            caption: <span>VoiceRecordingPlayer without progress bar</span>,
+            alt: 'Image of the VoiceRecordingPlayer missing progress bar',
+        },
+    ]}
+/>
+
+
+## Audio player controller
+
+The API to control the voice recording player is provided by <GHComponentLink text='useAudioController' path='/Attachment/hooks/useAudioController.tsx'/>. The `VoiceRecordingPlayer` makes use of the following memoized functions:
+
+1. `togglePlay` - to start and stop playing the record
+2. `seek` - to change audio progress
+3. `increasePlaybackRate` - to change the playback rate in the round-robin manner
+
+Besides the above actions, the controller maintains state in the following form:
+
+1. `isPlaying` - boolean flag informing the UI, whether the reproduction is in flight
+2. `progress` - numeric fractional value between 0 and 100 representing the progress of the audio reproduction
+3. `playbackRate` - currently selected playback rate value
+4. `secondsElapsed` - related to `progress`, but representing the amount of seconds elapsed since the audio beginning
```

Comment:
This seems to be an internal controller. I'm not sure if we should document it here. Perhaps it belongs to a cookbook, as I don't otherwise see how one can use or override this controller.

Reply:
True, I had doubts. I will remove it.